include/megaapi.h

```diff
@@ -12529,6 +12684,103 @@ class MegaApi
+         * Provided the error code is API_OK, the URL is available in the MegaBackgroundMediaUpload via its getUploadURL().
+         *
+         * Call this function just once (per file) to find out the URL to upload to, and upload all the pieces to the same
+         * URL.   If errors are encountered and the operation must be restarted from scratch, then a new URL should be requested.
+         * A new URL could specify a different upload server for example.
+         *
+         * @param fullFileSize The size of the file
+         * @param state A pointer to the MegaBackgroundMediaUpload object tracking this upload
+         * @param listener The MegaRequestListener to be called back with the result
+         */
+        void backgroundMediaUploadRequestUploadURL(int64_t fullFileSize, MegaBackgroundMediaUpload* state, MegaRequestListener *listener);
+
+        /**
+         * @brief Create the node after completing the background upload of the file.
+         *
+         * Call this function after completing the background upload of all the file data
+         * The node representing the file will be created in the cloud, with all the suitable
+         * attributes and file attributes attached.
+         * The result of the request is signalled by the listener onRequestFinsish callback with TYPE_COMPLETE_BACKGROUND_UPLOAD.
+         *
+         * @param state The MegaBackgroundMediaUpload object tracking this upload
+         * @param utf8Name The leaf name of the file, utf-8 encoded
+         * @param parent The folder node under which this new file should appear
+         * @param fingerprint  The fingerprint for the uploaded file (use MegaApi::getFingerprint to generate this)
+         * @param fingerprintoriginal If the file uploaded is modified from the original,
+         *        pass the fingerprint of the original file here, otherwise NULL.
+         * @param binaryUploadToken The N binary bytes of the token returned from the file upload (of the last portion). N=36 currently.
+         * @param listener The MegaRequestListener to be called back with the result
+         */
+        bool backgroundMediaUploadComplete(MegaBackgroundMediaUpload* state, const char* utf8Name, MegaNode *parent,
+            const char* fingerprint, const char* fingerprintoriginal, std::string* binaryUploadToken, MegaRequestListener *listener);
+
+        /**
+         * @brief Call this to enable the library to attach media info attributes
+         *
+         * Those attributes enable the web browser to know a file is a video, and
+         * play it with the correct codec.
+         *
+         * @return True if the library is ready, otherwise the request for media translation
+         *         data is sent to MEGA.  In that case, call again later or use a listener
+         *         to find out when the translation data is available.
+         */
+        bool ensureMediaInfo();
```

Comment:
Good spotting on the procresult().  That code was from the transition from no version to having a version at the front, and tried to cope with either from the API.  Now we can be sure a version is expected first, I'll update it.
About isMediaInfoReady, if we arrange it like that we could end up with two independent parts of a client app both requesting the `mc` data, and actually fetch it twice even though we only need one copy.   Should we be concerned about that?  thanks

Reply:
Well, fetch the same data twice for no strong reason is definitely not the way to go :) But usually apps are aware of having queued a request, and they wait for the result. That's the approach we've been using for more than 5 years. That's why I didn't like too much this approach. However, now that apps can listen to the `onEvent(EVENT_MEDIA_INFO_READY)`, I think we are all good.
Also, I tried to convert the `ensureMediaInfo()` into a request, but it would require a bunch of changes and philosophy, since the command is launched from different places internally, as well as in result of app's calls, and it started to be complicated to decouple it. On top of that, it's specific to the `MediaInfo` and that's already legacy code for this feature, so I preferred to live with this `call and silently wait`, but it's not a desired approach, since it leaves black holes where things may start to fail